os/drivers/pwm.c

```diff
@@ -445,27 +445,40 @@ static int pwm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 
 	case PWMIOC_SETCHARACTERISTICS: {
 		FAR const struct pwm_info_s *info = (FAR const struct pwm_info_s *)((uintptr_t)arg);
-		DEBUGASSERT(info != NULL && lower->ops->start != NULL);
+		DEBUGASSERT(info != NULL);
 
 #ifdef CONFIG_PWM_PULSECOUNT
 		pwmvdbg("PWMIOC_SETCHARACTERISTICS frequency: %d duty: %08x count: %d started: %d\n", info->frequency, info->duty, info->count, upper->started);
 #else
 		pwmvdbg("PWMIOC_SETCHARACTERISTICS frequency: %d duty: %08x started: %d\n", info->frequency, info->duty, upper->started);
 #endif
-
-		/* Save the pulse train characteristics */
-
-		memcpy(&upper->info, info, sizeof(struct pwm_info_s));
-
 		/* If PWM is already running, then re-start it with the new characteristics */
-
 		if (upper->started) {
+			DEBUGASSERT(lower->ops->stop != NULL);
+			ret = lower->ops->stop(lower);
```

Comment:
PWMIOC_SETCHARACTERISTICS sets the characteristics of the next pulsed output, but it might not change the characteristics on the fly. So for safety, it firstly stops the current pulse, and then, after changing the characteristics such as duration and period, the pulse will be restarted below.